smartsim/_core/mli/infrastructure/control/event_listener.py

```diff
@@ -0,0 +1,337 @@
+
+
+# pylint: enable=unused-import
+# pylint: enable=import-error
+# isort: on
+
+import argparse
+import multiprocessing as mp
+import os
+import sys
+import typing as t
+
+from smartsim._core.entrypoints.service import Service
+from smartsim._core.mli.comm.channel.dragon_channel import DragonCommChannel
+from smartsim._core.mli.comm.channel.dragon_util import create_local
+from smartsim._core.mli.infrastructure.storage.backbone_feature_store import (
+    BackboneFeatureStore,
+    EventBase,
+    EventCategory,
+    EventConsumer,
+    OnCreateConsumer,
+    OnRemoveConsumer,
+)
+from smartsim.error.errors import SmartSimError
+from smartsim.log import get_logger
+
+logger = get_logger(__name__)
+
+
+class ConsumerRegistrationListener(Service):
+    """A long-running service that listens for events of a specific type
+    and executes the appropriate event handler."""
+
+    def __init__(
+        self,
+        backbone: BackboneFeatureStore,
+        timeout: float,
+        batch_timeout: float,
+        as_service: bool = False,
+        cooldown: int = 0,
+        health_check_frequency: float = 60.0,
+    ) -> None:
+        """Initialize the EventListener.
+
+        :param backbone: The backbone feature store
+        :param timeout: Maximum time (in seconds) to allow a single recv request to wait
+        :param batch_timeout: Maximum time (in seconds) to allow a batch of receives to
+         continue to build
+        :param filters: Filters specifying the message types to handle
+        :param as_service: Specifies run-once or run-until-complete behavior of service
+        :param cooldown: Number of seconds to wait before shutting down after
+        shutdown criteria are met
+        """
+        super().__init__(
+            as_service, cooldown, health_check_frequency=health_check_frequency
+        )
+
+        self._timeout = timeout
+        """ Maximum time (in seconds) to allow a single recv request to wait"""
+
+        self._batch_timeout = batch_timeout
+        """Maximum time (in seconds) to allow a batch of receives to
+         continue to build"""
+
+        self._consumer: t.Optional[EventConsumer] = None
+        """The event consumer that handles receiving events"""
+
+        self._backbone = backbone
+        """A standalone, system-created feature store used to share internal
+        information among MLI components"""
+
+    def _on_start(self) -> None:
+        """Called on initial entry into Service `execute` event loop before
+        `_on_iteration` is invoked."""
+        super()._on_start()
+        self._create_eventing()
+
+    def _on_shutdown(self) -> None:
+        """Release dragon resources. Called immediately after exiting
+        the main event loop during automatic shutdown."""
+        super()._on_shutdown()
+
+        # unregister this listener in the backbone
+        self._backbone.pop(BackboneFeatureStore.MLI_BACKEND_CONSUMER)
+
+        # TODO: need the channel to be cleaned up
```

Comment:
Reminder that theres a TODO here!

Reply:
adding ticket for cleaning up all resources for next sprint